### PR TITLE
Refactored to NestJs, with test and layer architecture

### PR DESCRIPTION
**Pull request by learningSherban**
Added strategy for handling different currency providers

### CODE REVIEW

File: src/modules/currency/api/privatbank.api.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        this.authorization = "";
+    }
+
+    async getRate(targetCurrency: string): Promise<number | null> {
+        const config: AxiosRequestConfig = {
+            method: "get",
+            url: `/p24api/pubinfo?json&exchange&coursid=5`,
+        };
+        const res = (await this.request(config)) as AxiosResponse;
+        console.log(res.data.find((el: PrivatBankCurrency) => el.ccy === targetCurrency));
```
Comment (grynevychs) [inline src/modules/currency/api/privatbank.api.ts]: Looks like a debug statement to be removed

File: src/modules/currency/currency.strategy.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+@injectable()
+export class StrategySelector implements IStrategySelector {
+    constructor(
+        @inject(INVERSIFY_TYPES.IPrivatBankStrategy) private privatBankStrategy: ICurrencyStrategy,
+        @inject(INVERSIFY_TYPES.IOpenExchangeStrategy) private openExchangeStrategy: ICurrencyStrategy
+    ) {}
+
+    select(strategyName: CurrencyStrategies): ICurrencyStrategy {
+        console.log("strategyName", strategyName)
```
Comment (grynevychs) [inline src/modules/currency/currency.strategy.ts]: Log to be removed

File: src/modules/currency/currency.strategy.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+}
+
+@injectable()
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export class PrivatBankStrategy implements ICurrencyStrategy {
+    constructor(@inject(INVERSIFY_TYPES.IPrivatBankApi) private privatBankApi: IExchangeApi) {}
+
+    async getRate(targetCurrency: string) {
+        console.log("PRIV ST")
```
Comment (grynevychs) [inline src/modules/currency/currency.strategy.ts]: Log to be removed

File: src/modules/currency/currency.strategy.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+}
+
+@injectable()
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export class OpenExchangeStrategy {
+    constructor(@inject(INVERSIFY_TYPES.IOpenExchangeApi) private openExchangeApi: IExchangeApi) {}
+
+    async getRate(targetCurrency: string) {
+        console.log("OPEN ST")
```
Comment (grynevychs) [inline src/modules/currency/currency.strategy.ts]: Log to be removed

File: src/app.controller.spec.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { AppController } from './app.controller';
+import { AppService } from './app.service';
+
```
Comment (skelemen-skelar) [inline src/app.controller.spec.ts]: good, that you've added tests in this iteration. Ideally they should be moved to separate PR for HW#3 because it looks strange

File: src/modules/currency/api/openexchange.api.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { HttpService } from "@nestjs/axios";
+import { Injectable } from "@nestjs/common";
+import { BaseAxiosApi } from "src/common/api/base-axios.api";
+import { IExchangeApi } from "../interfaces/exchange-api.interface";
+import { firstValueFrom } from "rxjs";
+import { DotenvConfigService } from "src/config/dotenv/config.service";
+
+@Injectable()
+export class OpenExchangeApi extends BaseAxiosApi implements IExchangeApi {
```
Comment (skelemen-skelar) [inline src/modules/currency/api/openexchange.api.ts]: OpenExchangeApi extends BaseAxiosApi that implemets IExchangeApi that extends BaseAxiosApi
I suppose BaseAxiosApi and IExchangeApi can be combined to one abstract class with methods and properties from both

File: src/config/database/postgres/config.service.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import { Injectable } from "@nestjs/common";
```
Comment (skelemen-skelar) [inline src/config/database/postgres/config.service.ts]: interesting approach to configs but it can be significantly simplified.
here's sample I quickly created as [draft PR](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-learningSherban/pull/5)

Comment (skelemen-skelar) [reply src/config/database/postgres/config.service.ts]: it may be not 100% working because I wasn't testing it but it's exact approach we're using on real project

File: src/jobs/consumer/currency-rate-cron/currency-rate-cron.consumer.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+@Injectable()
+export class CurrencyRateCronService {
+    constructor(
+        private readonly subscriptionService: SubscriptionService,
+        private readonly currencyService: CurrencyService,
+        private readonly mailService: MailerService
+    ) {}
+
+    @Cron("* * * * * *")
```
Comment (skelemen-skelar) [inline src/jobs/consumer/currency-rate-cron/currency-rate-cron.consumer.ts]: doesn't it trigger request every second?

File: src/modules/currency/currency.chain.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    private providers: IExchangeApi[];
+
+    constructor(
+        private readonly privatbankApi: PrivatBankApi,
+        private readonly openexchangeApi: OpenExchangeApi
+    ) {
+        this.providers = [this.openexchangeApi, this.privatbankApi];
+    }
+
+    async getCurrencyData(baseCurrency: string): Promise<number> {
```
Comment (skelemen-skelar) [inline src/modules/currency/currency.chain.ts]: good :)

File: src/modules/subscription/entities/subscription.entity.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+export class Subscription implements ISubscription {
+    @PrimaryGeneratedColumn("uuid")
+    id: string;
+
+    @Column({ type: "varchar" })
+    email: string;
+
+    @CreateDateColumn({ name: "created_at", type: "timestamp" })
+    createdAt: Date;
+    @UpdateDateColumn({ name: "updated_at", type: "timestamp" })
```
Comment (skelemen-skelar) [inline src/modules/subscription/entities/subscription.entity.ts]: you can read about this https://www.npmjs.com/package/typeorm-naming-strategies#usage, not to do snake_case naming manually

File: src/modules/currency/spec/currency.chain.spec.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { OpenExchangeApi } from '../api/openexchange.api';
+import { PrivatBankApi } from '../api/privatbank.api';
+import { CurrencyChain } from '../currency.chain';
+import { BaseAxiosApi } from '@common/api/base-axios.api';
```
Comment (skelemen-skelar) [inline src/modules/currency/spec/currency.chain.spec.ts]: unused

**Review by skelemen-skelar — COMMENTED** [review]
Overall all tasks are done - that's good. But commiting postgres DB to github is crazy))) - please add it to .gitignore + you don't have [migrations](https://orkhan.gitbook.io/typeorm/docs/migrations) for db. Also there are some inaccuracies in OOP approach, unused code (prettier should be configured) and a bit of clutter all over the code.

**Comment by learningSherban** [comment]
@skelemen-skelar  I agree with the comments. I will finish all the moments after my next PR, which will already have a microservice approach. Now it is necessary to push code in the main branch so that there would be an opportunity to make the next PR
